vowpalwabbit/reductions/epsilon_decay.cc

```diff
@@ -0,0 +1,247 @@
+  ep_fts.epsilon = decayed_epsilon(active_iter->update_count);
+  base.predict(examples, active_iter->get_model_idx());
+}
+
+void learn(epsilon_decay_data& data, VW::LEARNER::multi_learner& base, multi_ex& examples)
+{
+  CB::cb_class logged{};
+  uint64_t labelled_action = 0;
+  const auto it =
+      std::find_if(examples.begin(), examples.end(), [](example* item) { return !item->l.cb.costs.empty(); });
+  if (it != examples.end())
+  {
+    logged = (*it)->l.cb.costs[0];
+    labelled_action = std::distance(examples.begin(), it);
+  }
+
+  const float r = -logged.cost;
+  // Process each model, then update the upper/lower bounds for each model
+  for (auto config_iter = data.scored_configs.begin(); config_iter != data.scored_configs.end(); ++config_iter)
+  {
+    // Update the scoring of all configs
+    // Only call if learn calls predict is set
+    auto& ep_fts = examples[0]->_reduction_features.template get<VW::cb_explore_adf::greedy::reduction_features>();
+    ep_fts.epsilon = decayed_epsilon(config_iter->update_count);
+    if (!base.learn_returns_prediction) { base.predict(examples, config_iter->get_model_idx()); }
+    base.learn(examples, config_iter->get_model_idx());
+    for (const auto& a_s : examples[0]->pred.a_s)
+    {
+      if (a_s.action == labelled_action)
+      {
+        const float w = (logged.probability > 0) ? a_s.score / logged.probability : 0;
+        config_iter->update_bounds(w, r);
+        break;
+      }
+    }
+  }
+
+  // If the lower bound of a model exceeds the upperbound of the champion, migrate the new model as
+  // the new champion.
+  auto champion_iter = data.scored_configs.rbegin();
+  auto end_iter = data.scored_configs.rend();
+  uint64_t model_count = data.scored_configs.size();
+  for (auto candidate_iter = champion_iter + 1; candidate_iter != end_iter; ++candidate_iter)
+  {
+    if (candidate_iter->get_lower_bound() > champion_iter->get_upper_bound())
+    {
+      auto n_iter = swap_models(candidate_iter, champion_iter, end_iter);
+      reset_models(n_iter, end_iter, data.weights, data.epsilon_decay_alpha, data.epsilon_decay_tau, model_count);
+      break;
+    }
+  }
+
+  // check if any model counts are higher than the champion. If so, shift the model
+  // back to the beginning of the list and reset its counts
+  auto model_idx = model_count - 1;
+  for (auto candidate_iter = champion_iter + 1; candidate_iter != end_iter; ++candidate_iter, --model_idx)
+  {
+    if (candidate_iter->update_count > data.min_scope &&
+        candidate_iter->update_count >
+            (std::pow(champion_iter->update_count, static_cast<float>(model_idx) / model_count)))
+    {
+      auto n_iter = swap_models(candidate_iter + 1, candidate_iter, end_iter);
+      reset_models(n_iter, end_iter, data.weights, data.epsilon_decay_alpha, data.epsilon_decay_tau, model_count);
+      break;
+    }
+  }
+}
+
+void save_load_epsilon_decay(epsilon_decay_data& epsilon_decay, io_buf& io, bool read, bool text)
```

Comment:
I dont think this is necessary. If all.save_resume is false (--predict_only_model is turned on with some existing model) then the weights should not be updated, and a new model should not be saved, since the current model being used is only for prediction.